Refactor of Consent Mode js to external file through webpack.

benbowler opened the pull request:
## Summary

<!-- Please reference the issue this PR addresses in the following list. -->
Addresses issue:

- #8384

## Relevant technical choices

Update and addressed all elements in the IB.

## PR Author Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 5.2 and PHP 7.4.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [x] Run the code.
- [x] Ensure the acceptance criteria are satisfied.
- [x] Reassess the implementation with the IB.
- [x] Ensure no unrelated changes are included.
- [x] Ensure CI checks pass.
- [x] Check Storybook where applicable.
- [x] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [ ] Ensure the PR has the correct target branch.
- [ ] Double-check that the PR is okay to be merged.
- [ ] Ensure the corresponding issue has a ZenHub release assigned.
- [ ] Add a changelog message to the issue.

zutigrm commented on assets/js/consent-mode/consent-mode.js:
```diff
@@ -0,0 +1,79 @@
+				const consentParameters = {};
+				let hasConsentParameters = false;
+				Object.keys( event.detail ).forEach( ( category ) => {
+					if ( window._googlesitekitConsentCategoryMap[ category ] ) {
```
```suggestion
                                        if ( global._googlesitekitConsentCategoryMap[ category ] ) {
```
Also in the rest of the file, `global` is prefered over `window`

zutigrm commented on assets/js/consent-mode/consent-mode.test.js:
```diff
@@ -0,0 +1,71 @@
+/**
+ * Consent Mode tests
+ *
+ * Site Kit by Google, Copyright 2025 Google LLC
```
```suggestion
 * Site Kit by Google, Copyright 2024 Google LLC
```
Still a bit early xD

zutigrm commented on assets/js/consent-mode/consent-mode.test.js:
```diff
@@ -0,0 +1,71 @@
+		} );
+	} );
+
+	it( 'should call gtag with the correct parameters when wp_consent_type_defined event is triggered', () => {
```
This seems to be a duplicate, since the previous test is the same

benbowler replied:
This test tests the route through the `wp_consent_type_defined` event:

https://github.com/google/site-kit-wp/blob/452957960539e9ae61275c93c6f2d2612a0f3c86/assets/js/consent-mode/consent-mode.js#L68-L71

Where as the first event is testing the route via the `wp_listen_for_consent_change` event:

https://github.com/google/site-kit-wp/blob/452957960539e9ae61275c93c6f2d2612a0f3c86/assets/js/consent-mode/consent-mode.js#L17-L42

zutigrm replied:
Ah I see, thanks

zutigrm commented on includes/Core/Consent_Mode/Consent_Mode.php:
```diff
@@ -73,9 +84,12 @@ public function register() {
 			);
+
+			// Register the consent mode script.
+			add_action( 'wp_enqueue_scripts', fn () => $this->enqueue_script(), 30 );
```
Is there a specific reason to use this priority of `30`? I did not notice change in enqueue order when switching to the default value

zutigrm reviewed (changes requested):
Thanks @benbowler Looks good overall. I just left you a few comments

aaemnnosttv commented on includes/Core/Consent_Mode/Consent_Mode.php:
```diff
@@ -88,12 +102,31 @@ function () use ( $consent_mode_enabled ) {
+	 * @since n.e.x.t
+	 */
+	protected function enqueue_script() {
+		list( $filename ) = Manifest::get( 'consent-mode' );
```
In normal usage, assets don't need to touch the manifest directly. Please see my comment on https://github.com/google/site-kit-wp/issues/8955#issuecomment-2206167344

benbowler replied:
Thanks @aaemnnosttv, yes that's right. [I've corrected this](https://github.com/google/site-kit-wp/pull/8893/commits/16d175dd1b1cb2279a8a06e75fbe0c28093a05b6)

aaemnnosttv commented on webpack/basicModules.config.js:
```diff
@@ -39,6 +39,8 @@ module.exports = ( mode ) => ( {
 		// Analytics advanced tracking script to be injected in the frontend.
 		'analytics-advanced-tracking':
 			'./assets/js/analytics-advanced-tracking.js',
+		// Consent Mode script.
+		'consent-mode': './assets/js/consent-mode/consent-mode.js',
```
This isn't really the right group for the script since "basic modules" are essentially unoptimized.

The intention of this issue is to extract the source to its own file/module to allow for tooling, testing, and the use of modern language features and leave the transpilation to widely-supported JS syntax to Babel. Since we don't have any such modules in the plugin yet, we likely need to create a new Webpack config for generating such output as the Babel config for SK entrypoints in wp-admin won't be as broad in terms of compatibility we want to have on the frontend.

benbowler replied:
Thanks @aaemnnosttv, yes that's right. [I've corrected this](https://github.com/google/site-kit-wp/pull/8893/commits/16d175dd1b1cb2279a8a06e75fbe0c28093a05b6)

benbowler replied:
In the end I renamed the "Event Providers" build config to be "Frontend Scripts" instead and updated the webpack config to use `@babel/preset-env` to transpile the scripts to support any browser which is not dead and has over 0.25% worldwide market share according to browserlist in the commit f9ea3e3ea2cac0d8a24acc2e0c21df84bbbb5cfa

aaemnnosttv replied:
This entry is a duplicate that still needs to be removed.

zutigrm reviewed (approved):
Thanks @benbowler that's all from me, assigning it to the @aaemnnosttv

benbowler commented:
I went through this with @aaemnnosttv, in our 1:1 today and I've updated this PR based on some comments there, notably:

1. I added the prefix to the conversion event provider files [here](https://github.com/google/site-kit-wp/pull/8893/files#diff-ef3970845733be83dc9cdab6b71a1caa74212f381b96d96c24dfe473bbc1aceb).
2. I confirmed the `webpack/frontendModules.config.js` config does minify the code in production.
3. I renamed the config to `webpack/frontendModules.config.js` from `webpack/frontendScripts.config.js`.

Moving to MR.

aaemnnosttv commented on assets/js/consent-mode/consent-mode.js:
```diff
@@ -0,0 +1,77 @@
+( function () {
+	document.addEventListener(
+		'wp_listen_for_consent_change',
+		function ( event ) {
```
How about we extract this to a named function as well (similar to `updateGrantedConsent` below) to avoid extra indentation and wrapping from longer lines?

aaemnnosttv commented on webpack/frontendModules.config.js:
```diff
@@ -35,6 +35,9 @@ const {
 module.exports = ( mode ) => ( {
 	entry: {
+		// Consent Mode
+		'consent-mode': './assets/js/consent-mode/consent-mode.js',
```
Noting here that the source file does not need to be prefixed but the URL reference to the built output in `dist/assets/js` would be.

```suggestion
		'googlesitekit-consent-mode': './assets/js/consent-mode/consent-mode.js',
```

aaemnnosttv commented on webpack.config.js:
```diff
@@ -42,8 +42,8 @@ function* webpackConfig( env, argv ) {
 
-	// Build conversion event provider files.
-	yield conversionEventProviders( mode );
+	// Build modules that will be used on the frontend, that require wider browser support.
+	yield frontendModules( mode );
```
Perfect 👍

aaemnnosttv commented on assets/js/consent-mode/consent-mode.js:
```diff
@@ -0,0 +1,77 @@
+/**
```
Just to be clear: this file name does not need to change.

aaemnnosttv reviewed (changes requested):
Thanks @benbowler – just a few things left to address. I've added a few other suggestions as well.

aaemnnosttv reviewed (approved):
LGTM, thanks @benbowler !
